### PR TITLE
[IMP] http: allow to handle HTTP and JSON requests on the same endpoint

### PR DESCRIPTION
**Pull request by std-odoo**
Purpose
=======
Be able to send/receive Facebook messages in Discuss.

Specifications
==============
Routing type
----------------
Actually, there are only 2 routing types, `http` and `json`.
But, in some case, we might need to manage standard HTTP
requests and JSON requests, on the same endpoint.

That's the case with the Facebook webhook. They first make a
simple HTTP request to check if the endpoint is a valid webhook
endpoint and then, the events are sent in a JSON request.

That's why we need a new routing type.

Customize the header of the thread window
---------------------------
We want to be able to customize the thread window, so we can have a blue header for the Facebook channel.

`anonymous_name` field
------------------------
In livechat, we have an "anonymous name" on the mail channel.
This field is useful because we do not have one partner per visitor.

In social, we also need this field (as we do not have a partner per
Facebook user) so we move it to mail.

Task-2124457
See odoo/enterprise/pull/8481
See odoo/iap-apps/pull/166
See odoo/upgrade/pull/956

### CODE REVIEW

File: odoo/http.py [hunk 1]
```diff
@@ -318,7 +318,7 @@ def _handle_exception(self, exception):
 
     def _call_function(self, *args, **kwargs):
         request = self
-        if self.endpoint.routing['type'] != self._request_type:
+        if self.endpoint.routing['type'] != self._request_type and self.endpoint.routing['type'] != 'raw':
```
Comment (awa-odoo) [inline odoo/http.py]: Several comments on this:

- Why not making "type" accept an array instead then?
  So you could do `type=['json', 'http']`.
  "raw" is pretty meaningless and if we ever add a third type of controller method, then it's broken.
- The feature should correctly handle method parameters whether it's called in 'http' or 'json', like other endpoint types.
- If you are going to add such a generic feature, you will need to write a few tests for it.
  (Check that all types work, parameters are correctly parsed, ...)

I understand you need to be able to handle both types, but this commit feels more like a "dirty fix" than a fully working new request type.

File: addons/mail/static/src/xml/discuss.xml [hunk 2]
```diff
@@ -135,7 +135,7 @@
         </t>
         <t t-if="!disableAddThread || !empty">
             <div class="o_mail_sidebar_title">
-                <h4 t-att-class="type == 'multi_user_channel' ? 'o_mail_open_channels' : ''">
+                <h4 t-att-class="type == 'multi_user_channel' ? 'o_mail_open_channels' : '' + 'd-flex flex-row justify-content-start align-items-stretch'">
```
Comment (awa-odoo) [inline addons/mail/static/src/xml/discuss.xml]: Commit is missing meaningful comment.

File: odoo/http.py (hunk 1 above)
Comment (std-odoo) [reply odoo/http.py]: @awa-odoo It can not be an array, because the "http handler" is a class, and is chosen thanks to the routing type...
So, when the type is `raw`, we take the HTTP handler but we have no restriction on the "request" type

You think I should add a new handler and not using the basic HTTP handler ?

Comment (std-odoo) [reply odoo/http.py]: @awa-odoo 
Using a list instead of a new parameter might need more code (because all the actual code use a string and not a list), and I prefer to write less code as possible in `http.py` (to avoid to break existing stuff).

Actually, the handler (the object used to parse the HTTP headers, and to call controllers) chose automatically the right handler (JSON or HTTP). And, they check if the handler is correct, regarding the routing type.
The `raw` type just ignore this verification. I think it's the simpler way to do it.

I can think about a better way to do it, but I don't know how

But I agree on the point that I should add test about it.

Comment (std-odoo) [reply odoo/http.py]: @awa-odoo 
But in the spec, est wanted to have the setting for all media, I'll ask her want she prefer

File: addons/mail/static/src/xml/discuss.xml (hunk 2 above)
Comment (awa-odoo) [inline addons/mail/static/src/xml/discuss.xml]: Drop this commit and the icon for Facebook.

It's not aligned with other channel names and looks strange.

File: addons/test_website/tests/test_controller_args.py [hunk 3]
```diff
@@ -29,3 +33,19 @@ def test_crawl_args(self):
         req = self.url_open('/ignore_args/kw?a=valueA&b=valueB')
         self.assertEqual(req.status_code, 200)
         self.assertEqual(req.json(), {'a': 'valueA', 'kw': {'b': 'valueB'}})
+
+    def test_route_type(self):
+        HOST = 'http://127.0.0.1'
```
Comment (awa-odoo) [inline addons/test_website/tests/test_controller_args.py]: Why don't you use `self.url_open` like other tests?

Comment (std-odoo) [reply addons/test_website/tests/test_controller_args.py]: @awa-odoo Because I can only perform `GET` requests with "url_open" (and bellow I need to perform `POST` requests)

File: addons/test_website/tests/test_controller_args.py [hunk 4]
```diff
@@ -29,3 +31,18 @@ def test_crawl_args(self):
         req = self.url_open('/ignore_args/kw?a=valueA&b=valueB')
         self.assertEqual(req.status_code, 200)
         self.assertEqual(req.json(), {'a': 'valueA', 'kw': {'b': 'valueB'}})
+
+    def test_route_type(self):
+        with MockRequest(self.env):
```
Comment (awa-odoo) [inline addons/test_website/tests/test_controller_args.py]: Do you need to use MockRequest here as you're not using the request object?

File: addons/test_website/controllers/main.py [hunk 5]
```diff
@@ -119,3 +119,7 @@ def get_post_method(self, **kw):
     @http.route(['/get_post_nomultilang'], type='http', auth="public", methods=['GET', 'POST'], website=True, multilang=False, sitemap=False)
     def get_post_method_no_multilang(self, **kw):
         return request.make_response('get_post_nomultilang')
+
+    @http.route(['/test_route_1'], type='*', auth="public", methods=['GET', 'POST'], csrf=False)
+    def test_route_1(self, **kw):
```
Comment (awa-odoo) [inline addons/test_website/controllers/main.py]: Maybe improve naming to `test_route_json_http` ?

And also add a small docstring what it does / returns.

File: addons/test_website/controllers/main.py [hunk 6]
```diff
@@ -119,3 +119,7 @@ def get_post_method(self, **kw):
...
+        return str(kw) + str(request.jsonrequest if hasattr(request, 'jsonrequest') else '')
```
Comment (awa-odoo) [inline addons/test_website/controllers/main.py]: Can't you do a if/else based on the hasattr instead?
Would be easier to understand that you return one parameter or the other based on the type.

Also, is there no cleaner way to check the request type?

(I assume that "kw" is empty when you access the route with a json type).

File: addons/test_website/tests/test_controller_args.py (hunk 4 above)
Comment (std-odoo) [reply addons/test_website/tests/test_controller_args.py]: no, indeed :)

**Comment by awa-odoo** [comment]
@std-odoo 

Side note: you can rewrite the PR description to match your commit message since it also contains a single commit after all.

Right now it's a bit confusing.

**Comment by std-odoo** [comment]
@awa-odoo Done :)

**Comment by tde-banana-odoo** [comment]
@odony Could be great to have your final word on it, enterprise is ready for being launched.

**Comment by tde-banana-odoo** [comment]
@robodoo r+ rebase-ff

File: odoo/http.py [hunk 7]
```diff
@@ -521,7 +521,7 @@ def response_wrap(*args, **kw):
                     _logger.info("<function %s.%s> called ignoring args %s" % (f.__module__, f.__name__, ', '.join(ignored)))
 
             response = f(*args, **kw)
-            if isinstance(response, Response) or f.routing_type == 'json':
+            if isinstance(response, Response) or f.routing_type in ['json', '*']:
```
Comment (odony) [inline odoo/http.py]: This means the case of an HTTP request won't be handled exactly the same with `type='*'` as with `type='http'`, e.g. next lines won't apply. Could we keep it functionally equivalent with something like this instead? 

```suggestion
            if isinstance(response, Response) or isinstance(request._get_current_object(), JsonRequest):
```

ref: https://werkzeug.palletsprojects.com/local/#werkzeug.local.LocalProxy._get_current_object

File: addons/mail/models/mail_channel.py [hunk 8]
```diff
@@ -430,7 +431,7 @@ def _send_guidelines(self, partners):
         return True
 
     def _update_moderation_email(self, emails, status):
-        """ This method adds emails into either white or black of the channel list of emails 
+        """ This method adds emails into either white or black of the channel list of emails
```
Comment (odony) [inline addons/mail/models/mail_channel.py]: Please refrain from linting lines that your commit does not change, as it adds noise and messes up blame history ;-)

File: addons/test_website/tests/test_controller_args.py [hunk 9]
```diff
@@ -29,3 +31,17 @@ def test_crawl_args(self):
...
+        self.assertEqual(response.text, "{'blabla': 'qsd'}")
+
+        response = self.url_open('/test_route_json_http', data=b"blabla=qsd", headers={'Content-type': 'application/x-www-form-urlencoded'})
+
+        self.assertEqual(response.status_code, 200)
+        self.assertEqual(response.text, "{'blabla': 'qsd'}")
+
+        response = self.url_open('/test_route_json_http', data=b'{"a": "blabla"}', headers={'Content-type': 'application/json'})
+        self.assertEqual(response.status_code, 200)
+        self.assertTrue("{'a': 'blabla'}" in response.text)
```
Comment (odony) [inline addons/test_website/tests/test_controller_args.py]: I guess this comment was lost in a push-force, but I was asking for some improvements of the test logic here:
- test the actual JSON response with `response.json['result']`
- perhaps have the route return structured data in the JSON case, to show a more meaningful use case (here it is almost identical to the HTTP case, which I found confusing while reading)

File: addons/test_website/tests/test_controller_args.py [hunk 10]
```diff
@@ -29,3 +31,17 @@ def test_crawl_args(self):
         req = self.url_open('/ignore_args/kw?a=valueA&b=valueB')
         self.assertEqual(req.status_code, 200)
         self.assertEqual(req.json(), {'a': 'valueA', 'kw': {'b': 'valueB'}})
+
+    def test_route_type(self):
+        response = self.url_open('/test_route_json_http?blabla=qsd')
+        self.assertEqual(response.status_code, 200)
+        self.assertEqual(response.text, "{'blabla': 'qsd'}")
+
+        response = self.url_open('/test_route_json_http', data=b"blabla=qsd", headers={'Content-type': 'application/x-www-form-urlencoded'})
```
Comment (odony) [inline addons/test_website/tests/test_controller_args.py]: For new code, I like to keep lines under 100-120 chars, it's less annoying to view on github split mode (no wrapping), and nicer on the eyes of readers. Thanks! ;-)

**Comment by tde-banana-odoo** [comment]
@robodoo r-

File: odoo/http.py (hunk 7 above)
Comment (std-odoo) [reply odoo/http.py]: @odony Hi :)
It doesn't always work, it seems that sometimes the "current object" is a python dict
See
https://runbot.odoo.com/runbot/build/1951024
_tmp_screenshot

Comment (std-odoo) [reply odoo/http.py]: IMO it's safer to stay like this, to not impact existing routes

File: addons/test_website/tests/test_controller_args.py [hunk 11]
```diff
@@ -1,5 +1,7 @@
 # Part of Odoo. See LICENSE file for full copyright and licensing details.
+
 import odoo.tests
+from odoo.addons.website.tools import MockRequest
```
Comment (awa-odoo) [inline addons/test_website/tests/test_controller_args.py]: Is this still used ?

File: addons/test_website/tests/test_controller_args.py [hunk 12]
```diff
@@ -29,3 +31,25 @@ def test_crawl_args(self):
         req = self.url_open('/ignore_args/kw?a=valueA&b=valueB')
         self.assertEqual(req.status_code, 200)
         self.assertEqual(req.json(), {'a': 'valueA', 'kw': {'b': 'valueB'}})
+
+    def test_route_type(self):
+        response = self.url_open('/test_route_json_http?blabla=qsd')
+        self.assertEqual(response.status_code, 200)
+        self.assertEqual(response.text, "{'blabla': 'qsd'}")
```
Comment (dbeguin) [inline addons/test_website/tests/test_controller_args.py]: I'm sure you can rename this into `?url_param=param_name` :smile:

File: addons/mail/static/src/xml/abstract_thread_window.xml [hunk 13]
```diff
@@ -5,7 +5,7 @@
         @param {mail.AbstractThreadWindow} widget
     -->
     <t t-name="mail.AbstractThreadWindow">
-        <div class="o_thread_window o_in_home_menu" t-att-data-thread-id="widget.getID()">
+        <div t-attf-class="o_thread_window o_in_home_menu o_thread_window_#{widget._thread &amp;&amp; widget._thread._type}" t-att-data-thread-id="widget.getID()">
```
[the remark this answers is not in the document]
Comment (std-odoo) [reply addons/mail/static/src/xml/abstract_thread_window.xml]: @dbeguin 
In the way we want :p In example, I use the CSS class to change to color of the header

[the remark this answers is not in the document]
Comment (std-odoo) [reply addons/mail/static/src/xml/abstract_thread_window.xml]: but ok, I update the commit message :)

Comment (odony) [inline addons/mail/static/src/xml/abstract_thread_window.xml:8]: more readable and supported on the JS fwk side IIRC
```suggestion
        <div t-attf-class="o_thread_window o_in_home_menu o_thread_window_#{widget._thread and widget._thread._type}" t-att-data-thread-id="widget.getID()">
```

File: odoo/http.py (hunk 7 above)
Comment (odony) [reply odoo/http.py:530]: The reason for this error is that `MockRequest` does not simulate a real request: https://github.com/odoo/odoo/blob/13.0/addons/website/tools.py#L96-L116
It's a bit of a pity to be held back in the core code because of implementation details of the test system.

What about introducing a `request.type` attribute then? It would be a very small change (example below), and would allow to test `request.type == 'json'` here instead of the suggested `isinstance()`, to the same effect.

And then you only have to adapt the few tests that call a json route with MockRequest, and pass `type='json'` explicitly.
It will also make a slightly faster and more elegant way of testing the request type than doing `hasattr(request, 'jsonrequest')` in other places.

The idea is something like this:
```diff
diff --git addons/website/tools.py addons/website/tools.py
index 3639776706fe..bbdb352d4859 100644
--- addons/website/tools.py
+++ addons/website/tools.py
@@ -112,6 +112,7 @@ class MockRequest(object):
                 'sale_order_id': kw.get('sale_order_id'),
             },
             'website': kw.get('website'),
+            'type': kw.get('type', 'http'),
         })
         odoo.http._request_stack.push(self.request)
 
diff --git odoo/http.py odoo/http.py
index 50874bfbd50f..337d8fbb3939 100644
--- odoo/http.py
+++ odoo/http.py
@@ -203,6 +203,7 @@ class WebRequest(object):
         self.endpoint = None
         self.endpoint_arguments = None
         self.auth_method = None
+        self.type = 'http'
         self._cr = None
         self._uid = None
         self._context = None
@@ -577,6 +578,7 @@ class JsonRequest(WebRequest):
     def __init__(self, *args):
         super(JsonRequest, self).__init__(*args)
 
+        self.type = 'json'
         self.params = {}
 
         args = self.httprequest.args

```

Comment (std-odoo) [reply odoo/http.py]: @odony But so, that means that we can call only one "route type" in a `MockRequest` namespace.
Is it ok ?

Comment (std-odoo) [reply odoo/http.py]: (and there's already the attribute `_request_type` :) no need to init it in the `__init__`)

Comment (std-odoo) [reply odoo/http.py]: (and also, pay attention that the routing type is different than the request type, the request type can be "http" or "json" and the routing type can be "http", "json", or "*", it add a little confusion imo, but for a temporary solution it's ok)

File: addons/test_website/tests/test_controller_args.py [hunk 14]
```diff
@@ -29,3 +30,25 @@ def test_crawl_args(self):
         req = self.url_open('/ignore_args/kw?a=valueA&b=valueB')
         self.assertEqual(req.status_code, 200)
         self.assertEqual(req.json(), {'a': 'valueA', 'kw': {'b': 'valueB'}})
+
+    def test_route_type(self):
```
Comment (awa-odoo) [inline addons/test_website/tests/test_controller_args.py:34]: Maybe rename to test_route_type_all and add a bit of docstring?

Comment (awa-odoo) [reply addons/test_website/tests/test_controller_args.py:34]: Also, side note for commit message:
"Actually" is a fake friend in English, its meaning is close to "En fait" in French.
If you want "Actuellement", "Currently" is more suitable.

File: addons/mail/models/mail_message.py [hunk 15]
```diff
@@ -1100,6 +1100,7 @@ def message_format(self):
                 'is_discussion': message_sudo.subtype_id.id == com_id,
                 'subtype_description': message_sudo.subtype_id.description,
                 'is_notification': vals['message_type'] == 'user_notification',
+                'channel_types': list(message_sudo.mapped('channel_ids.channel_type')),
```
Comment (awa-odoo) [inline addons/mail/models/mail_message.py:1103]: Why do you have to specify "list()" here?

Comment (awa-odoo) [reply addons/mail/models/mail_message.py:1103]: Also, I don't know discuss in details but I'm not sure why this information has to be in every single messages.
Surely when you display the message there is a way to know in which channel and channel type you are to condition some behavior right?

Seems like it adds complexity / potential overhead where it doesn't need to.

Comment (std-odoo) [reply addons/mail/models/mail_message.py:1103]: @awa-odoo 
In JS, in mail message, I only have access to the ID (int) of the channel, so if I want to know if the message is inside a social channel, I'll need to do a RPC call :/

I use this information to show the `anonymous_name` of the channel instead of the name of the partner ("public partner") if the message is inside a social channel.
(In livechat the situation is not the same because the name of the channel is the name of the partner...-

The other solution is to inherit from `'mail.message' -> message_format` in social and to add the field (like I did before, but people didn't like it).

The last solution would be to "include" `MailManager` , and to change the method `_makeMessage` to return a new message type (with the social information) if we are in a social channel. But, I'll need to create one more JS class (and it will probably be broken after the discuss refactoring).

Comment (std-odoo) [reply addons/mail/models/mail_message.py:1103]: (sorry for the very long message :/ )

File: addons/mail/static/src/xml/abstract_thread_window.xml [hunk 16]
```diff
@@ -5,7 +5,7 @@
         @param {mail.AbstractThreadWindow} widget
     -->
     <t t-name="mail.AbstractThreadWindow">
-        <div class="o_thread_window o_in_home_menu" t-att-data-thread-id="widget.getID()">
+        <div t-attf-class="o_thread_window o_in_home_menu o_thread_window_#{widget._thread and widget._thread._type}" t-att-data-thread-id="widget.getID()">
```
Comment (seb-odoo) [inline addons/mail/static/src/xml/abstract_thread_window.xml:8]: Avoid automatically crafted classes because those are very hard to maintain (impossible to find it back in the code from the full name). It's preferable to add conditions and explicit names.

Using private variables in a template is also not a good sign.

File: addons/mail/static/src/xml/thread.xml [hunk 17]
```diff
@@ -312,9 +312,10 @@
                         <t t-esc="message.getDisplayedAuthor()"/>
                     </strong>
                     <strong t-elif="message.hasEmailFrom()">
-                        <a class="text-muted" t-attf-href="mailto:#{message.getEmailFrom()}?subject=Re: #{message.hasSubject() ? message.getSubject() : ''}">
+                        <a t-if="message.shouldRedirectToAuthor()" class="text-muted" t-attf-href="mailto:#{message.getEmailFrom()}?subject=Re: #{message.hasSubject() ? message.getSubject() : ''}">
```
Comment (seb-odoo) [inline addons/mail/static/src/xml/thread.xml:315]: Is this change necessary? What is the case it is trying to cover?

`hasEmailFrom` is on an `elif` of which the corresponding `if` is `hasAuthor`, and `shouldRedirectToAuthor` will always return `true` if `hasAuthor` is `false` (through `_isMyselfAuthor`).

File: odoo/http.py [hunk 18]
```diff
@@ -528,7 +528,7 @@ def response_wrap(*args, **kw):
                     _logger.info("<function %s.%s> called ignoring args %s" % (f.__module__, f.__name__, ', '.join(ignored)))
 
             response = f(*args, **kw)
-            if isinstance(response, Response) or f.routing_type == 'json':
+            if isinstance(response, Response) or request._request_type == 'json':
```
Comment (JKE-be) [inline odoo/http.py:531]: Suggestion, (to validate with @odony or @antonylesuisse)

But if I'm not wrong, the main purpose of this feature type="*" is to support call in http on a json controller.
So basically, be able todo a $.get("/route.json") and receive a str(dict)

We should probably handle the case where your controller return a dict (json case) and that you call the route in http on a route *.

```diff
diff --git odoo/http.py odoo/http.py
index 85388a8589d..edc861580d0 100644
--- odoo/http.py
+++ odoo/http.py
@@ -534,6 +534,9 @@ def route(route=None, **kw):
             if isinstance(response, (bytes, str)):
                 return Response(response)
 
+            if request.endpoint.routing['type'] == "*" and isinstance(response, dict) and request._request_type == 'http':
+                return Response(str(response))
+
             if isinstance(response, werkzeug.exceptions.HTTPException):

```

And from this way, avoid the exception when you call in http a controller '*' that return a dict
`TypeError: 'dict' object is not callable - - -
`



Example:

```python
    @http.route('/controller/star', type='*', auth="public")
    def website_star(self, **kwargs):
        return {
            'feature1': kwargs.get('value1'),
            'feature2': kwargs.get('value2'),
            'version': odoo.service.common.exp_version()
        }
```

```js
>> odoo.__DEBUG__.services['web.ajax'].jsonRpc('/controller/star', 'call', {value1:'val1', value2:'val2'}).then(o=>console.log(o))
```
> {"jsonrpc": "2.0", "id": 918268530, "result": {"feature1": "val1", "feature2": "val2", "version": {"server_version": "13.4alpha1", "server_version_info": [13, 4, 0, "alpha", 1, ""], "server_serie": "13.4", "protocol_version": 1}}}

```js
>> $.get('/controller/star?value1=val1&value2=val2').then(o=>console.log(o));
```

> {'feature1': 'val1', 'feature2': 'val2', 'version': {'server_version': '13.4alpha1', 'server_version_info': (13, 4, 0, 'alpha', 1, ''), 'server_serie': '13.4', 'protocol_version': 1}}

File: addons/mail/static/src/xml/thread.xml (hunk 17 above)
Comment (std-odoo) [reply addons/mail/static/src/xml/thread.xml:315]: @seb-odoo yes, because in social we have an `email_from` (with the Facebook user info), but we don't want to be able to click on the contact name in Discuss

Comment (seb-odoo) [reply addons/mail/static/src/xml/thread.xml:315]: Ok for the reasoning, but the method `shouldRedirectToAuthor` still does not make sense if `hasAuthor` is `false`, it should never be called in that context. It's actually currently returning an incorrect value in that case, because it returns `true` but there is no author set, so to which author it should redirect?!

I think you need to find a condition that specifically work for your Facebook case (do you know it is a Facebook message?), and it should likely be an override in `social`.

Comment (std-odoo) [reply addons/mail/static/src/xml/thread.xml:315]: @seb-odoo It that case, we do not redirect to `mailto: <email_from>`, I can add a new method `shouldRedirectToEmailFrom`, but I think it's cleaner to use the same method.

(if an author exist, `shouldRedirectToAuthor` tell us if we redirect to the `res.partner`, otherwise, it tell us if we should open the email in the default mail client when clicking on the partner name))

Hope it's cear

Comment (seb-odoo) [reply addons/mail/static/src/xml/thread.xml:315]: Having one method doing 2 things in 2 different contexts is not a good idea.

I would prefer a new method such as `shouldRedirectToEmailFrom`, but with another name, because it is not a "redirect" when it's an email, but it's a "mailto link". So maybe just `hasMailtoLink`.

Note that is it out of scope of this PR, but in livechat the mailto link is displayed even if no email address is set/it is invalid, so that method will also help avoid that case in the future.

File: odoo/http.py (hunk 18 above)
Comment (std-odoo) [reply odoo/http.py:531]: @JKE-be Indeed, good idea :)

I just call `json.dumps` instead of `str` because single quotes are not valid in a JSON :)

File: addons/mail/static/src/xml/abstract_thread_window.xml (hunk 16 above)
Comment (std-odoo) [reply addons/mail/static/src/xml/abstract_thread_window.xml]: @seb-odoo I added the type in attribute, so I can select it in CSS without crafting classes ;)

File: addons/mail/models/mail_channel.py [hunk 19]
```diff
@@ -105,6 +105,7 @@ def default_get(self, fields):
     image_128 = fields.Image("Image", max_width=128, max_height=128, default=_get_default_image)
     is_subscribed = fields.Boolean(
         'Is Subscribed', compute='_compute_is_subscribed')
+    anonymous_name = fields.Char('Anonymous Name')
```
Comment (awa-odoo) [inline addons/mail/models/mail_channel.py:108]: Maybe a good opportunity to add a help attribute explaining this field usage in a small sentence?

File: addons/test_website/controllers/main.py [hunk 20]
```diff
@@ -120,6 +120,19 @@ def get_post_method(self, **kw):
...
         return request.make_response('get_post_nomultilang')
 
+    @http.route(['/test_route_json_http'], type='*', auth="public", methods=['GET', 'POST'], csrf=False)
+    def test_route_json_http(self, **kw):
+        if hasattr(request, 'jsonrequest'):
+            return request.jsonrequest
+        return str(kw)
+
+    @http.route(['/test_route_json_http_dict'], type='*', auth="public", methods=['GET'], csrf=False)
+    def test_route_json_http_dict(self):
```
Comment (awa-odoo) [inline addons/test_website/controllers/main.py:130]: The method naming is a bit obscure and the docstring surely doesn't help.

File: addons/mail/static/src/js/models/messages/message.js [hunk 21]
```diff
@@ -59,6 +59,22 @@ var Message =  AbstractMessage.extend(Mixins.EventDispatcherMixin, ServicesMixin
...
+    /**
+     * Get anonymous name of the message
+     *
+     * @override
+     * @return {string}
+     */
+    getAnonymousName: function () {
+        let originalChannelId = this.getOriginChannelID();
+        if (originalChannelId >= 0) {
+            let anonymousName = this.call('mail_service', 'getThread', originalChannelId).getAnonymousName();
```
Comment (awa-odoo) [inline addons/mail/static/src/js/models/messages/message.js:71]: Maybe implement the default `_t("Anonymous")` value in the channel's `getAnonymousName` definition as well?

That way the behavior is similar if any other feature needs to get the anonymous name of the channel.
(It will also allow you to return the method result right away and simplify the code).

**Comment by rousseldenis** [comment]
@std-odoo I think this should be changed according to:

https://github.com/odoo/odoo/pull/78857

and the introduction of Dispatcher

**Comment by std-odoo** [comment]
oops this is an old PR I forgot to close, sorry
